GameRoom.ts

```diff
@@ -1,500 +1,637 @@
+              if (
+                TileSetController.generateField(
+                  fields,
+                  this.state.boardLayout,
+                  false
+                )
+              ) {
+                console.log('generated random Layout');
+              } else {
+                console.error('failed to generate random Layout');
+              }
             }
-
-            // re-enable figure old game
-            const pObj: PhysicsObjectState = this.state.physicsState.objects[player.figureId];
-            pObj.setDisabled(false);
-
-            // register new sessionId to old player object
-            player.clientId = client.id;
-            player.hasLeft = false;
+            try {
+            } catch (err) {
+              console.error(
+                'an error occurred while trying to generate the Field',
+                err
+              );
+            }
+          },
+          (reason: any) => {
+            console.error("failed, couldn't get TileSet.fields", reason);
+          }
+        );
+      },
+      (reason: any) => {
+        console.error("failed, couldn't get TileSet", reason);
+      }
+    );
+
+    this.state.physicsState.setBroadcastCallback(this.broadcast.bind(this));
+    this.state.physicsState.addDice();
+    this.state.physicsState.setOnDiceThrow(
+      ((num: number) => {
+        if (this.state.action === Actions[Actions.ROLL]) {
+          this.state.nextAction(this);
         }
-
-        player.isConnected = true;
-        this.broadcast(MessageType.JOIN_MESSAGE, { type: MessageType.JOIN_MESSAGE, message: joinedMsg });
-        return undefined;
+      }).bind(this)
+    );
+
+    this.onMessage(MessageType.CHAT_MESSAGE, this.onChatMessage.bind(this));
+    this.onMessage(MessageType.JOIN_MESSAGE, this.onJoinMessage.bind(this));
+    this.onMessage(MessageType.GAME_MESSAGE, this.onGameMessage.bind(this));
+    this.onMessage(MessageType.PLAYER_MESSAGE, this.onPlayerMessage.bind(this));
+    this.onMessage(
+      MessageType.PHYSICS_MESSAGE,
+      this.onPhysicsMessage.bind(this)
+    );
+    this.onMessage(MessageType.DEBUG_COMMAND, this.onDebugMessage.bind(this));
+    this.onMessage(MessageType.LEFT_MESSAGE, this.onLeftMessage.bind(this));
+    this.onMessage(MessageType.OTHER, this.onOtherMessage.bind(this));
+    this.onMessage(
+      MessageType.REFRESH_COMMAND,
+      this.onRefreshMessage.bind(this)
+    );
+    this.onMessage(
+      MessageType.ACHIEVEMENT_MESSAGE,
+      this.onAchievementMessage.bind(this)
+    );
+    this.onMessage(MessageType.ITEM_MESSAGE, this.onItemMessage.bind(this));
+    this.onMessage(MessageType.CHAT_COMMAND, this.onChatCommand.bind(this));
+
+    return undefined;
+  }
+
+  async onDispose(): Promise<void | Promise<any>> {
+    console.log(`[onDispose] Destructing physicsState`);
+
+    const tileSet = await TileSetController.getTileSetById(
+      this.metadata.tileSetId
+    );
+    await GameController.saveGameLog(
+      new Game(
```

Comment:
If the Game object is generated here, is it used anywhere other than to build the log? if so could we rename it to more meaningfully display its nature as a logging stata structure? liks: `GameLog`